src/App.js

```diff
@@ -25,44 +24,98 @@ const generateSquares = () => {
 };
 
+// Creating 2D array
+const reshapSquare = (arr) => {
+  const newArr = [];
+  while (arr.length) newArr.push(arr.splice(0, 3));
+  return newArr;
+};
+
 const App = () => {
-  // This starts state off as a 2D array of JS objects with
-  // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
+  const [nextPlayer, setNextPlayer] = useState(true);
+  const [gameWinner, setWinner] = useState(null);
+
+  const currentPlayer = nextPlayer ? PLAYER_1 : PLAYER_2;
+
+  const changeSquare = (id) => {
+    if (gameWinner !== null) {
+      return;
+    }
+
+    let flattenArray = squares.flat();
+    changedSquares = flattenArray.map((square) => {
+      if (square.id === id && square.value === '') {
+        square.value = currentPlayer;
+        console.log(currentPlayer, 'currentPlayer');
+      }
 
-  // Wave 2
-  // You will need to create a method to change the square
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+      return square;
+    });
+
+    setNextPlayer(!nextPlayer);
+    setSquares(reshapSquare(changedSquares));
+    setWinner(checkForWinner);
+  };
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let i = 0;
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
   };
 
   const resetGame = () => {
     // Complete in Wave 4
+    setSquares(generateSquares());
+    setNextPlayer(PLAYER_1);
+    // setSquares(reshapSquare(changedSquares));
+    setWinner(null);
+    console.log('reset');
   };
 
   return (
     <div className="App">
       <header className="App-header">
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        <h2>The winner is {gameWinner} </h2>
+        <h3> Current Player is{currentPlayer}</h3>
```

Comment:
Minor note: for better user experience, add a space:
```suggestion
        <h3> Current Player is {currentPlayer}</h3>
```